### PR TITLE
Upgrade a chart

### LINKED ISSUE #352: upgrade a chart
We can upgrade a chart.
image

There are two steps involved : 
- [ ] First is to saved the query string ( as we do for a saved config inside vault ) inside a specific values of the helm chart. onyxia.userDefinedValues.   
- [ ] In a second step we can create a button and linked it to the upgrade/install api calls. 

For that we need to save the fixed values somewhere (maybe in a custom values). 
Onyxia-web must fill all the automatic values then fixed values are apply ( fixed values overwrite automatic values )

### PR DESCRIPTION
**Pull request by LoicMnz**
Resolve https://github.com/InseeFrLab/onyxia-web/issues/352

### CODE REVIEW

File: src/core/adapters/officialOnyxiaApiClient.ts [hunk 1]
```diff
@@ -559,11 +559,13 @@ export function createOfficialOnyxiaApiClient(params: {
...
 
             const launchPackage = id<OnyxiaApiClient["launchPackage"]>(
-                async ({ catalogId, packageName, options, isDryRun }) => {
-                    const { serviceId } = getServiceId({
-                        packageName,
-                        "randomK8sSubdomain": getRandomK8sSubdomain()
-                    });
+                async ({ catalogId, packageName, options, isDryRun, name }) => {
+                    const serviceId =
+                        name ||
```
Comment (garronej) [inline src/core/adapters/officialOnyxiaApiClient.ts]: Là il convien d'utiliser `??` a la place de `||`

File: src/core/usecases/restorableLaunchPackageConfigs.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+};
+
+export type RestorableLaunchPackageConfig = {
+    name: string;
+    catalogId: string;
+    packageName: string;
+    formFieldsValueDifferentFromDefault: FormFieldValue[];
+};
+
+export const name = "restorableLaunchPackageConfig";
```
Comment (garronej) [inline src/core/usecases/restorableLaunchPackageConfigs.ts]: `runningPackageConfigs`

File: src/core/usecases/restorableLaunchPackageConfigs.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                "value": JSON.stringify(
+                    getState().restorableLaunchPackageConfig
+                        .restorableLaunchPackageConfigs
+                )
+            })
+        )
+};
+
+export const thunks = {
+    "saveRestorableLaunchPackageConfig":
```
Comment (garronej) [inline src/core/usecases/restorableLaunchPackageConfigs.ts]: `saveRunningPackageConfig`

File: src/core/usecases/restorableLaunchPackageConfigs.ts [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            }
+
+            dispatch(
+                actions.restorableLaunchPackageConfigsaved({
+                    restorableLaunchPackageConfig
+                })
+            );
+            await dispatch(privateThunks.syncWithUserConfig());
+        },
+    "deleterestorableLaunchPackageConfig":
```
Comment (garronej) [inline src/core/usecases/restorableLaunchPackageConfigs.ts]: `deleteRunningPackageConfig`

File: src/core/usecases/restorableLaunchPackageConfigs.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+            dispatch(
+                actions.restorableLaunchPackageConfigDeleted({
+                    serviceId
+                })
+            );
+            await dispatch(privateThunks.syncWithUserConfig());
+        },
+    /** Pure */
+    "isrestorableLaunchPackageConfigInStore":
```
Comment (garronej) [inline src/core/usecases/restorableLaunchPackageConfigs.ts]: `isRunningPackageConfigInStore`

File: src/core/usecases/userConfigs.ts [hunk 6]
```diff
@@ -29,6 +29,7 @@ export type UserConfigs = Id<
         githubPersonalAccessToken: string | null;
         doDisplayMySecretsUseInServiceDialog: boolean;
         bookmarkedServiceConfigurationStr: string | null;
+        launchServicesConfigurationStr: string | null;
```
Comment (garronej) [inline src/core/usecases/userConfigs.ts]: `runningPackagesConfigurationStr`

File: src/ui/components/pages/MyServices/MyServicesCards/MyServicesCard/MyServicesCard.tsx [hunk 7]
```diff
@@ -30,6 +30,7 @@ export type Props = {
     onRequestDelete: (() => void) | undefined;
     onRequestShowPostInstallInstructions: (() => void) | undefined;
     onRequestShowEnv: () => void;
+    onRequestUpgrade: () => void;
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServicesCards/MyServicesCard/MyServicesCard.tsx]:
```ts
/** undefined if not upgradable */
onRequestUpgrade: (()=> void) | undefined;
```

File: src/ui/components/pages/MyServices/MyServicesCards/MyServicesCards.tsx [hunk 8]
```diff
@@ -71,6 +81,10 @@ export const MyServicesCards = memo((props: Props) => {
         onRequestDelete({ serviceId })
     );
 
+    const onRequestUpgradeFactory = useCallbackFactory(([params]: [UpgradeParams]) =>
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServicesCards/MyServicesCards.tsx]: The memoization isn't working here. 
The `params` object ref changes every render.  

```ts
const onRequestUpgradeFactory = useCallbackFactory(([catalogIdd, packageName, name]: [string | undefined, string, string])=> 
  onRequestUpgrade({ catalogId, packageName, name })
);
```

File: src/ui/components/pages/MyServices/MyServices.tsx [hunk 9]
```diff
@@ -40,28 +40,31 @@ export function MyServices(props: Props) {
...
+    const { restorableContainerConfigs } = useCoreState(
+        selectors.restorableLaunchPackageConfig.restorableContainerConfigs
+    );
 
     const isRunningServicesUpdating = useCoreState(
         state => state.runningService.isUpdating
     );
-
+    const { isLaunching } = useCoreState(selectors.launcher.isLaunching);
+    const isLoading = isLaunching || isRunningServicesUpdating;
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServices.tsx]: There is an oportunity for a closure here:  
```typescript 
const { isLoading } = (function useClosure(){

       const { isLaunching } = useCoreState(selectors.launcher.isLaunching);

       const isRunningServicesUpdating = useCoreState(
           state => state.runningService.isUpdating
        );

	const isLoading = isLaunching || isRunningServicesUpdating;

	return { isLoading };

})(); 

```

File: src/ui/components/pages/MyServices/MyServices.tsx [hunk 10]
```diff
@@ -199,10 +202,17 @@ export function MyServices(props: Props) {
                           "isOwned": rest.isOwned,
                           "ownerUsername": rest.isOwned ? undefined : rest.ownerUsername,
                           vaultTokenExpirationTime,
-                          s3TokenExpirationTime
+                          s3TokenExpirationTime,
+                          catalogId: restorableContainerConfigs.find(
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServices.tsx]: Property name are double quoted in this codebase

File: src/ui/components/pages/MyServices/MyServices.tsx [hunk 11]
```diff
@@ -244,21 +254,38 @@ export function MyServices(props: Props) {
...
+            await launcher.reset();
+            await launcher.initialize({
+                catalogId: params.catalogId!,
+                packageName: params.packageName!,
+                name: params.name,
+                formFieldsValueDifferentFromDefault
+            });
+
+            await launcher.launch();
+            runningService.update();
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServices.tsx]: This should be a thunk (core function). This is not the responsability of the UI to perform such complex logic.

**Comment by garronej** [comment]
Sorry for the late review, thank you for your work.

File: src/core/usecases/userConfigs.ts (hunk 6 above)
Comment (LoicMnz) [reply src/core/usecases/userConfigs.ts]: Done

File: src/ui/components/pages/MyServices/MyServicesCards/MyServicesCards.tsx (hunk 8 above)
Comment (LoicMnz) [reply src/ui/components/pages/MyServices/MyServicesCards/MyServicesCards.tsx]: Changed after move the upgradeService in core function. It remains just name.

**Comment by LoicMnz** [comment]
Thanks to you for the review

File: src/ui/components/pages/MyServices/MyServicesCards/MyServicesCard/MyServicesCard.tsx [hunk 12]
```diff
@@ -40,6 +42,7 @@ export type Props = {
     ownerUsername: string | undefined;
     vaultTokenExpirationTime: number | undefined;
     s3TokenExpirationTime: number | undefined;
+    isUpgradable: boolean;
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServicesCards/MyServicesCard/MyServicesCard.tsx]: I mean the condition that tells if the chart is upgradable can be deduced by looking if onRequestUpgrade() is `undefined` or not.  
`isUpgradable` is overdeterministic and introduce unecessary coupling ( !!isUpgradable => onRequestUpgrade !== undefined)

File: src/ui/components/pages/MyServices/MyServicesCards/MyServicesCards.tsx [hunk 13]
```diff
@@ -44,6 +46,7 @@ export type Props = {
         | undefined;
     catalogExplorerLink: Link;
     onRequestDelete(params: { serviceId: string }): void;
+    onRequestUpgrade(serviceName: string): void;
```
Comment (garronej) [inline src/ui/components/pages/MyServices/MyServicesCards/MyServicesCards.tsx]: !!!!!!!   
Please try to enforce consistency.
Just look at the above line.  
If you want to rename `serviceId` to `serviceName` fine. But do it in a separate PR and propagate the change everywhere.

File: src/core/usecases/runningPackageConfigs.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+
+export const selectors = (() => {
+    function restorableRunnigPackageConfigs(rootState: State) {
+        const { runningPackageConfigs } = rootState.runningPackageConfigs;
+
+        return runningPackageConfigs
+            .map(runningPackageConfig => {
+                return {
+                    runningPackageConfig
+                };
```
Comment (garronej) [inline src/core/usecases/runningPackageConfigs.ts]: Why wrapping here?  
It's okay to copy paste as a starting point but everything must be carefully reviewed.

File: src/core/usecases/runningPackageConfigs.ts [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+            })
+            .reverse();
```
Comment (garronej) [inline src/core/usecases/runningPackageConfigs.ts]: Does it make sence to sort here?  
I think not.

File: src/core/usecases/runningPackageConfigs.ts [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+
+function areSameRunningPackageConfig(
+    runningPackageConfiguration1: RunningPackageConfig,
+    runningPackageConfiguration2: RunningPackageConfig
+): boolean {
+    return runningPackageConfiguration1.name === runningPackageConfiguration2.name;
+}
+
+export const selectors = (() => {
+    function restorableRunnigPackageConfigs(rootState: State) {
```
Comment (garronej) [inline src/core/usecases/runningPackageConfigs.ts]: The thunk has been renamed but the selector hasn't.  
What's this selector? How does it differs from the internal state.  
If it's doesn't, just give it the same name.

File: src/core/usecases/runningService.ts [hunk 17]
```diff
@@ -135,6 +140,38 @@ export const thunks = {
...
+                if (state.isUpdating) {
+                    return;
+                }
+            }
+            const restorableRunningPackageConfigs =
+                runningPackageConfigSelectors.restorableRunnigPackageConfigs(getState());
+            const restorableRunningPackageConfig = restorableRunningPackageConfigs.find(
+                e => e.runningPackageConfig.name === serviceId
+            )?.runningPackageConfig;
+            assert(restorableRunningPackageConfig !== undefined);
```
Comment (garronej) [inline src/core/usecases/runningService.ts]: You didn't take the time to propagate your renaming.  
I have the context so I can figure out what's going on but it will be criptic for the next guy.

File: src/core/usecases/runningService.ts [hunk 18]
```diff
@@ -135,6 +140,38 @@ export const thunks = {
...
+
+            await dispatch(
+                launcherThunk.initialize({
+                    catalogId: restorableRunningPackageConfig.catalogId,
+                    packageName: restorableRunningPackageConfig.packageName,
+                    name: restorableRunningPackageConfig.name,
+                    formFieldsValueDifferentFromDefault:
+                        restorableRunningPackageConfig.formFieldsValueDifferentFromDefault
```
Comment (garronej) [inline src/core/usecases/runningService.ts]: We use double quote for object property in this codebase.  
You may disagree but it's important that the coding style be enforced throughout the codebase.

File: src/core/usecases/runningService.ts [hunk 19]
```diff
@@ -135,6 +140,38 @@ export const thunks = {
...
+                })
+            );
+            await dispatch(launcherThunk.launch());
+            dispatch(launcherThunk.reset());
+            dispatch(thunks.update());
```
Comment (garronej) [inline src/core/usecases/runningService.ts]: This is hacky.  
It's like simulating buttons clicks.  
This thunks has been desined specifically as button callbacks for the UI, not as functions that can be composed to acheive other behaviour.  

I'm going to approve this anyway...

**Comment by garronej** [comment]
Here is my reveiw.
